Automattic/newspack-plugin
# feat(salesforce): check duplicate site status and disable syncs if clone

### All Submissions:






### Changes proposed in this Pull Request:



This avoids potential bad syncs if a site with an active Salesforce connection is cloned for staging or development purposes.

Basically, in such a case, if a site with an active Salesforce connection is cloned, any active subscriptions will continue to generate recurring orders, which will then get synced to Salesforce. This could result in bad or duplicate data if both the live site and its clones are sending recurring donation data.

This PR prevents this from happening using the [same mechanism](https://github.com/woocommerce/-unused-woocommerce-subscriptions/blob/master/woocommerce-subscriptions.php#L1099) WooCommerce Subscriptions uses to put itself into [staging mode](https://woocommerce.com/document/subscriptions-handles-staging-sites/).

If you need a dev Salesforce account to test with, please DM me for credentials.

### How to test the changes in this Pull Request:

1. You need to simulate the conditions for a clone or staging site. To do so, you can either [set up WooCommerce + Subscriptions and a Salesforce integration on an actual live site](https://newspack.pub/support/reader-revenue/) and clone it, or on your local dev site, add this temporary filter somewhere in your code. This will create a mismatch between the URL reported by WC Subscriptions vs. WP's site URL. (You can add it to the same `class-webhooks.php` file if you need, or your theme's `functions.php`.)

```php
add_filter(
	'wc_subscriptions_site_url',
	function( $url ) {
			return 'https://foo';
	}
);
```

2. In Reader Revenue, set your Platform to Newspack and ensure you have a working Salesforce connection.
3. After this, make a donation on your site and wait a few minutes.
4. Go to WooCommerce > Status > Logs and look in the dropdown for a `webhooks-delivery` log for today. (If you don't see one, wait a bit longer and refresh.)
5. In the log, look for a webhook request with details matching your donation, and verify that the response was a 500 error like this:

```
[Response] => Array
	(
		[Code] => 500
		[Message] => Internal Server Error
		[Headers] => Requests_Utility_CaseInsensitiveDictionary Object
			(
				[data:protected] => Array
					(
						[access-control-allow-headers] => Authorization, X-WP-Nonce, Content-Disposition, Content-MD5, Content-Type
						[access-control-expose-headers] => X-WP-Total, X-WP-TotalPages, Link
						[allow] => POST, PUT, PATCH
						[content-type] => application/json; charset=UTF-8
						[date] => Wed, 05 Jan 2022 17:51:11 GMT
						[link] => <https://newspack-dkoo.ngrok.io/wp-json/>; rel="https://api.w.org/"
						[server] => nginx/1.16.0
						[x-content-type-options] => nosniff
						[x-distributor] => yes
						[x-powered-by] => PHP/7.4.1
						[x-robots-tag] => noindex
					)

			)

		[Body] => {"code":"newspack_salesforce_duplicate_site","message":"This site is a staging or development clone. Third-party syncs will not be processed.","data":null}
	)
```

6. In your connected Salesforce account, verify that the donation data was NOT synced.

### Other information:
